mypy incorrectly resolves names inside decorated methods

* Are you reporting a bug, or opening a feature request?

A bug

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.
```
from typing import Type

class A:
    pass

class B:
    @staticmethod
    def A() -> Type[A]:
        return A

assert A is B.A()
```

* What is the actual behavior/output?
```
$ mypy mypy_fail.py
mypy_fail.py:8: error: Function "mypy_fail.B.A" is not valid as a type
mypy_fail.py:8: note: Perhaps you need "Callable[...]" or a callback protocol?
Found 1 error in 1 file (checked 1 source file)
```

* What is the behavior/output you expect?

I expect this to be accepted - it works fine in python

Probably of interest: regular methods work like I expect. `classmethod` and `property` complain the same way, so I assume this has to do with decorators. For functions, I can't reproduce syntax like this that's also valid python (I get `UnboundLocalError`s), so that's as far as I got with debugging.

* What are the versions of mypy and Python you are using?
  Do you see the same issue after installing mypy from Git master?

mypy `0.750` and python 3.6/3.8. I also get it on git master.

* What are the mypy flags you are using? (For example --strict-optional)

No flags required

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -3949,7 +3949,9 @@ class SemanticAnalyzer(NodeVisitor[None],
         """
         assert self.statement
         line_diff = self.statement.line - node.line
-        if isinstance(node, Decorator):
+        if isinstance(node, OverloadedFuncDef) and isinstance(self.statement, FuncDef):
+            return node.name != self.statement.name and line_diff > 0
+        elif isinstance(node, Decorator) and not node.is_overload:
             return line_diff > len(node.original_decorators)
         else:
             return line_diff > 0

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -3953,16 +3953,20 @@ class SemanticAnalyzer(NodeVisitor[None],
         # The first branch handles reference an overloaded function variant inside itself,
         # this is a corner case where mypy technically deviates from runtime name resolution,
         # but it is fine because we want an overloaded function to be treated as a single unit.
-        if isinstance(node, OverloadedFuncDef) and isinstance(self.statement, FuncDef):
-            return self.is_overloaded_item(node, self.statement) and line_diff > 0
+        if self.is_overloaded_item(node, self.statement):
+            return False
         elif isinstance(node, Decorator) and not node.is_overload:
             return line_diff > len(node.original_decorators)
         else:
             return line_diff > 0
 
-    def is_overloaded_item(self, overload: OverloadedFuncDef, func: FuncDef) -> bool:
+    def is_overloaded_item(self, node: SymbolNode, statement: Statement) -> bool:
         """Check whehter the function belongs to the overloaded variants"""
-        return func in overload.items
+        if isinstance(node, OverloadedFuncDef) and isinstance(statement, FuncDef):
+            return (statement in
+                    [item.func if isinstance(item, Decorator) else item for item in node.items]
+                    or (node.impl is not None and statement is node.impl))
+        return False
 
     def is_defined_in_current_module(self, fullname: Optional[str]) -> bool:
         if fullname is None:

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -3942,9 +3942,17 @@ class SemanticAnalyzer(NodeVisitor[None],
                 or (isinstance(node, PlaceholderNode) and node.becomes_typeinfo))
 
     def is_textually_before_statement(self, node: SymbolNode) -> bool:
+        """Check if a node is defined textually before a statement
+
+        Note that decorated functions' line number are the same as
+        the topest decorator.
+        """
         assert self.statement
         line_diff = self.statement.line - node.line
-        return line_diff > 1 or (line_diff == 1 and not isinstance(node, Decorator))
+        if isinstance(node, Decorator):
+            return line_diff > len(node.original_decorators)
+        else:
+            return line_diff > 0
 
     def is_defined_in_current_module(self, fullname: Optional[str]) -> bool:
         if fullname is None:

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -3936,10 +3936,15 @@ class SemanticAnalyzer(NodeVisitor[None],
         #       caught.
         assert self.statement  # we are at class scope
         return (node is None
-                or node.line < self.statement.line
+                or self.is_textually_before_statement(node)
                 or not self.is_defined_in_current_module(node.fullname)
                 or isinstance(node, TypeInfo)
                 or (isinstance(node, PlaceholderNode) and node.becomes_typeinfo))
+
+    def is_textually_before_statement(self, node: SymbolNode) -> bool:
+        assert self.statement
+        line_diff = self.statement.line - node.line
+        return line_diff > 1 or (line_diff == 1 and not isinstance(node, Decorator))
 
     def is_defined_in_current_module(self, fullname: Optional[str]) -> bool:
         if fullname is None:

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -3963,9 +3963,12 @@ class SemanticAnalyzer(NodeVisitor[None],
     def is_overloaded_item(self, node: SymbolNode, statement: Statement) -> bool:
         """Check whehter the function belongs to the overloaded variants"""
         if isinstance(node, OverloadedFuncDef) and isinstance(statement, FuncDef):
-            return (statement in
-                    [item.func if isinstance(item, Decorator) else item for item in node.items]
-                    or (node.impl is not None and statement is node.impl))
+            in_items = statement in {item.func if isinstance(item, Decorator)
+                                     else item for item in node.items}
+            in_impl = (node.impl is not None and
+                      ((isinstance(node.impl, Decorator) and statement is node.impl.func)
+                       or statement is node.impl))
+            return in_items or in_impl
         return False
 
     def is_defined_in_current_module(self, fullname: Optional[str]) -> bool:

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -3942,7 +3942,7 @@ class SemanticAnalyzer(NodeVisitor[None],
                 or (isinstance(node, PlaceholderNode) and node.becomes_typeinfo))
 
     def is_textually_before_statement(self, node: SymbolNode) -> bool:
-        """Check if a node is defined textually before a statement
+        """Check if a node is defined textually before the current statement
 
         Note that decorated functions' line number are the same as
         the top decorator.

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -3945,7 +3945,7 @@ class SemanticAnalyzer(NodeVisitor[None],
         """Check if a node is defined textually before a statement
 
         Note that decorated functions' line number are the same as
-        the topest decorator.
+        the top decorator.
         """
         assert self.statement
         line_diff = self.statement.line - node.line

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -3949,12 +3949,20 @@ class SemanticAnalyzer(NodeVisitor[None],
         """
         assert self.statement
         line_diff = self.statement.line - node.line
+
+        # The first branch handles reference an overloaded function variant inside itself,
+        # this is a corner case where mypy technically deviates from runtime name resolution,
+        # but it is fine because we want an overloaded function to be treated as a single unit.
         if isinstance(node, OverloadedFuncDef) and isinstance(self.statement, FuncDef):
-            return node.name != self.statement.name and line_diff > 0
+            return self.is_overloaded_item(node, self.statement) and line_diff > 0
         elif isinstance(node, Decorator) and not node.is_overload:
             return line_diff > len(node.original_decorators)
         else:
             return line_diff > 0
+
+    def is_overloaded_item(self, overload: OverloadedFuncDef, func: FuncDef) -> bool:
+        """Check whehter the function belongs to the overloaded variants"""
+        return func in overload.items
 
     def is_defined_in_current_module(self, fullname: Optional[str]) -> bool:
         if fullname is None:

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -6605,8 +6605,7 @@ class A:
 
 class B:
     @staticmethod
-    def A() -> Type[A]:
-        return A
+    def A() -> Type[A]: ...
     @staticmethod
     def B() -> Type[A]:  # E: Function "__main__.B.A" is not valid as a type \
                          # N: Perhaps you need "Callable[...]" or a callback protocol?

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -6633,5 +6633,8 @@ class B:
     @classmethod
     @overload
     def A(cls, x: str) -> Type[A]: ...
+    @classmethod
+    def A(cls, x: object) -> Type[A]:
+        return A
 
 [builtins fixtures/classmethod.pyi]

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -6633,7 +6633,8 @@ class B:
     @overload
     def A(cls, x: str) -> Type[A]: ...
     @classmethod
-    def A(cls, x: object) -> Type[A]:
-        return A
+    def A(cls, x: object) -> Type[A]: ...
+    def B(cls, x: int) -> Type[A]: ...  # E: Function "__main__.B.A" is not valid as a type \
+                                        # N: Perhaps you need "Callable[...]" or a callback protocol?
 
 [builtins fixtures/classmethod.pyi]

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -6607,6 +6607,10 @@ class B:
     @staticmethod
     def A() -> Type[A]:
         return A
+    @staticmethod
+    def B() -> Type[A]:  # E: Function "__main__.B.A" is not valid as a type \
+                         # N: Perhaps you need "Callable[...]" or a callback protocol?
+        return A
 
 class C:
     @property
@@ -6615,3 +6619,19 @@ class C:
         return A
 
 [builtins fixtures/staticmethod.pyi]
+
+[case testRefMethodWithOverloadDecorator]
+from typing import Type, overload
+
+class A:
+    pass
+
+class B:
+    @classmethod
+    @overload
+    def A(cls, x: int) -> Type[A]: ...
+    @classmethod
+    @overload
+    def A(cls, x: str) -> Type[A]: ...
+
+[builtins fixtures/classmethod.pyi]

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -6596,3 +6596,17 @@ reveal_type(D() + "str")  # N: Revealed type is 'Any'
 
 reveal_type(0.5 + D1())  # N: Revealed type is 'Any'
 reveal_type(D1() + 0.5)  # N: Revealed type is '__main__.D1'
+
+[case testRefMethodWithDecorator]
+from typing import Type
+
+class A:
+    pass
+
+
+class B:
+    @staticmethod
+    def A() -> Type[A]:
+        return A
+
+[builtins fixtures/classmethod.pyi]

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -6603,10 +6603,15 @@ from typing import Type
 class A:
     pass
 
-
 class B:
     @staticmethod
     def A() -> Type[A]:
         return A
 
-[builtins fixtures/classmethod.pyi]
+class C:
+    @property
+    @staticmethod
+    def A() -> Type[A]:
+        return A
+
+[builtins fixtures/staticmethod.pyi]

--- a/test-data/unit/fixtures/staticmethod.pyi
+++ b/test-data/unit/fixtures/staticmethod.pyi
@@ -9,6 +9,7 @@ class type:
 class function: pass
 
 staticmethod = object() # Dummy definition.
+property = object()  # Dummy definition
 
 class int:
     @staticmethod

